File: string-to-integer-atoi/string-to-integer-atoi.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
    let result = 0, sign = 1, index = 0;
    const maxInteger = Math.pow(2, 31) - 1;
    const minInteger = -Math.pow(2, 31);
    
    // We need to discard any leading 0s and spaces from the beginning of input string
    while(index < s.length && s[index] === ' ') {
          index += 1;
    }
    
    if (index < s.length && s[index] === '+') {
        sign = 1;
        index += 1;
    } else if (index < s.length && s[index] === '-') {
        sign = -1;
        index += 1;
    }
    
    // Traverse string and stop if current item is not a digit
    while(index < s.length && !isNaN(s[index]) && s[index] !== ' ') {
        let digit = s[index] - '0';
        
        // Check underflow and overflow 
        if (result > Math.floor(maxInteger / 10) || (result === Math.floor(maxInteger / 10) && digit > maxInteger % 10)) {
            return sign === 1 ? maxInteger : minInteger;
        }
        
        result  = result * 10 + digit;
        index += 1;
    }
    
    return sign * result;
};

File: string-to-integer-atoi/string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int num = 0; // result
        int index = 0;
        
        if (s == null || s.length() == 0) {
            return 0;
        }
        
        // Skip white spaces
        while (index < s.length() && s.charAt(index) == ' ') {
            index++;
        }
        
        // Check for sign
        int sign = 1;
        if (index < s.length() && (s.charAt(index) == '+' || s.charAt(index) == '-')) {
            sign = s.charAt(index) == '+' ? 1 : -1;
            index++;
        }
        
        while (index < s.length() && Character.isDigit(s.charAt(index))) {
            int digit = s.charAt(index) - '0';
            if (num > (Integer.MAX_VALUE - digit) / 10) {  // Check for overflow
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            num = num * 10 + digit;
            index++;
        }
        
        return sign * num;
    }
}